### PR TITLE
feat(ci): enable host-asan tests for external repo presubmit builds

### PR DESCRIPTION
**Pull request by raramakr**
## Motivation

Host-asan CI currently only runs on `schedule` and `workflow_dispatch`. This enables it on `pull_request` events for callers that opt in, so that ASAN regressions are caught on the PR that introduces them rather than overnight.

GitHub issue: https://github.com/ROCm/TheRock/issues/7202

## Technical Details

Adds an `asan_presubmit` boolean input to `setup_multi_arch.yml`, plumbed through to `configure_multi_arch_ci.py` as `CIInputs.asan_presubmit`.

The host-asan gate now enables tests on these triggers:

| trigger | condition |
|---|---|
| `schedule` | always enabled |
| `workflow_dispatch` | always enabled |
| `push` | never enabled |
| `pull_request` | enabled for external repos that set the `asan_presubmit` opt-in |

The opt-in is declared explicitly by the caller rather than inferred from `build_stages`, so narrowing a build graph for unrelated reasons does not start consuming ASAN runner capacity.

Prerequisites, both already merged:
- https://github.com/ROCm/TheRock/pull/7600
- https://github.com/ROCm/TheRock/pull/7602

## Test Plan

Unit tests for the gate in `configure_multi_arch_ci_test.py`: the opt-in enables tests for an external repo, its absence leaves them unscheduled, and the opt-in alone does not enable them without an external repo.

End-to-end validation was run from rocm-systems against a TheRock branch carrying this change, driven incrementally so each failure could be attributed to a specific cause.

## Test Result

Unit tests: 953 passed, 2 skipped (950 before this change; the 3 added are the delta).

End-to-end, https://github.com/ROCm/rocm-systems/actions/runs/35140846158 confirms the pipeline reaches real ASAN results rather than failing on infrastructure:

- Host-asan tests are scheduled on the PR trigger and receive a sandbox runner.
- The sanity job runs `hip_check` to completion and reports a symbolized `AddressSanitizer: bad-free` in `libamdhip64.so.7` at `hsa_amd_memory_pool_free`, with both the allocation and free stacks resolved. This is a genuine finding, not an instrumentation artifact: ASAN's interceptors are installed and it tracked the 4 MB region across its lifetime.
- `hip-tests` discovers and runs roughly 4,900 tests, of which about 2,050 pass; the remaining failures are the same pre-existing `bad-free`.

Earlier runs in the sequence failed on infrastructure rather than test content, and each was fixed separately:
- ASAN runtime resolution after the layout change in #8077 — https://github.com/ROCm/TheRock/pull/8268
- The libraries baseline gate failing instead of skipping when `build_stages` narrows the graph — https://github.com/ROCm/TheRock/pull/8161
- Sanity-check preload — https://github.com/ROCm/TheRock/pull/8123

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/TheRock/blob/main/CONTRIBUTING.md.

### CODE REVIEW

**Review by dileepr1 — APPROVED** [review]
The change looks good.

A nit: 
Adding a unit test might be good to make sure that we land on the right runner.

File: build_tools/github_actions/configure_multi_arch_ci.py
Comment (ScottTodd) [inline build_tools/github_actions/configure_multi_arch_ci.py:1278]: I'm not confident in "External repos that set build_stages (e.g., rocm-systems presubmit) have already scoped the build enough to make sandbox runners affordable on every PR." as we still haven't completed the work to enable multi-arch CI (non-asan) in external repositories. Let's stick that landing first before we add more jobs.

* We currently set `hip-clr` to `level = 3` meaning test "self + all transitive consumers" here: https://github.com/ROCm/TheRock/blob/36503ee49f840e134983a7a97272a6fe43d0052a/test_tools/test_policies.toml#L21-L26
* I would like to set RCCL to level 2 (https://github.com/ROCm/TheRock/blob/main/docs/rfcs/RFC0013-Consumer-Based-Test-Selection.md), to build and test pytorch. That decision could be separate for asan and non-asan

I expect we'll want finer-grained control here, not just setting "every external repository runs asan all the time".

**Comment by ScottTodd** [comment]
Stepping back a bit, what coverage are we expecting from running tests on asan CI? Do we need to run both asan tests and non-asan tests on PRs? That seems like duplicate coverage that will double our runner usage.

Looking at the example workflow run https://github.com/ROCm/rocm-systems/actions/runs/33412744162/job/99564499807, that failed on the driver / GPU sanity check. Do ASan tests work? Can we look at some logs together? I looked around some more and I can't find useful asan build/test logs to analyze:
* nightly release gfx942: timeout fetching artifacts https://github.com/ROCm/rockrel/actions/runs/33469678529/job/99736933174
* nightly release gfx950: sanity check amdsmi failure https://github.com/ROCm/rockrel/actions/runs/33437861217/job/99638861879
* submodule bump skipped tests: https://github.com/ROCm/TheRock/actions/runs/33403337559

We shouldn't just enable workflows on presubmit without first validating them on manual workflow_dispatch, postsubmit, or nightly (in general presubmit should be a subset of postsubmit), and I'm not seeing enough evidence that these test are functional.

Meta: these changes are quite difficult to review in a timely manner without seeing the plan written down and more complete evidence of testing. I'm looking for detailed analysis like what was recently done on https://github.com/ROCm/rocm-libraries/pull/11439#issuecomment-5460371128 for example.

**Comment by ScottTodd** [comment]
Still looking for any logs of asan tests actually running and working (this is a prerequisite for enabling _any_ workflow on pull requests, unless there is explicit justification for why known-failing jobs are being added):
* https://github.com/ROCm/rocm-systems/actions/workflows/therock-multi-arch-ci-asan.yml?query=event%3Aworkflow_dispatch workflow has never been run with `workflow_dispatch`
* https://github.com/ROCm/rocm-systems/actions/runs/32524644456 and https://github.com/ROCm/rocm-systems/actions/runs/31716463837 did run tests but they failed
* https://github.com/ROCm/rocm-systems/actions/workflows/therock-multi-arch-ci-asan-nightly.yml has always failed (zero successful runs: https://github.com/ROCm/rocm-systems/actions/workflows/therock-multi-arch-ci-asan-nightly.yml?query=is%3Asuccess)

File: build_tools/github_actions/configure_multi_arch_ci.py [hunk 1]
```diff
@@ -1261,9 +1261,16 @@ def _expand_build_config_for_platform(
...
-            if not (ci_inputs.is_schedule or ci_inputs.is_workflow_dispatch):
+            # Run host-asan tests on nightly (schedule or workflow_dispatch) or
+            # for external repo builds that have narrowed the build graph.
+            # External repos that set build_stages (e.g., rocm-systems presubmit)
+            # have already scoped the build enough to make sandbox runners
+            # affordable on every PR.
+            if not (
+                ci_inputs.is_schedule
+                or ci_inputs.is_workflow_dispatch
+                or (ci_inputs.external_repo and ci_inputs.build_stages)
```
Comment (geomin12) [inline build_tools/github_actions/configure_multi_arch_ci.py:1272]: can we just specify presubmit? why build_stages?

**Review by geomin12 — COMMENTED** [review]
also as Scott mentioned, let's make sure it passes on presubmit before we enable!

**Comment by raramakr** [comment]
> Still looking for any logs of asan tests actually running and working (this is a prerequisite for enabling _any_ workflow on pull requests, unless there is explicit justification for why known-failing jobs are being added):
> 
> * https://github.com/ROCm/rocm-systems/actions/workflows/therock-multi-arch-ci-asan.yml?query=event%3Aworkflow_dispatch workflow has never been run with `workflow_dispatch`
> * https://github.com/ROCm/rocm-systems/actions/runs/32524644456 and https://github.com/ROCm/rocm-systems/actions/runs/31716463837 did run tests but they failed
> * https://github.com/ROCm/rocm-systems/actions/workflows/therock-multi-arch-ci-asan-nightly.yml has always failed (zero successful runs: https://github.com/ROCm/rocm-systems/actions/workflows/therock-multi-arch-ci-asan-nightly.yml?query=is%3Asuccess)


1) The gpu driver sanity failures are fixed in the original PR. Its due to ASAN Pre load issue. Will raise a separate PR for the same as part of splitting the original PR(https://github.com/ROCm/TheRock/pull/7311/)
2) Even after fixing the gpu driver sanity, the tests will still fail due to memory leak issue in each component(hip, rocr)

Will provide a detailed analysis on what is planned and  the detailed test result

File: build_tools/github_actions/configure_multi_arch_ci.py (hunk 1 above)
Comment (dileepr1) [reply build_tools/github_actions/configure_multi_arch_ci.py]: Added changes to use asan_presubmit here.

**Review by nunnikri — APPROVED** [review]
LGTM

File: build_tools/github_actions/configure_multi_arch_ci.py [hunk 2]
```diff
@@ -1290,13 +1297,22 @@ def _expand_build_config_for_platform(
...
+            # gate. Presubmit is declared explicitly rather than inferred from
+            # build_stages: scoping the build graph and opting into sandbox test
+            # runners on every PR are separate decisions, and a caller that
+            # narrows its build for unrelated reasons should not silently start
+            # consuming ASAN runner capacity.
+            if not (
+                ci_inputs.is_schedule
+                or ci_inputs.is_workflow_dispatch
+                or (ci_inputs.external_repo and ci_inputs.asan_presubmit)
+            ):
```
Comment (ScottTodd) [inline build_tools/github_actions/configure_multi_arch_ci.py:1311]: Please use human words for comments like this or review agent comments with care. This comment is particularly confusing as written and it mentions multiple irrelevant details.

For example:
```suggestion
            # host-asan tests are resource intensive so they run on the
            # following triggers:
            #   trigger           | conditions
            #   ----------------- | --------------
            #   schedule:         | always enabled
            #   workflow_dispatch | always enabled
            #   push              | never enabled
            #   pull_request      | enabled if 'asan_presubmit' opt-in is set
            if not (
                ci_inputs.is_schedule
                or ci_inputs.is_workflow_dispatch
                or (ci_inputs.external_repo and ci_inputs.asan_presubmit)
            ):
```

File: .github/workflows/setup_multi_arch.yml [hunk 3]
```diff
@@ -94,6 +94,13 @@ on:
         description: "Build native Linux packages (deb/rpm) and run their install tests."
         type: boolean
         default: true
+      asan_presubmit:
+        description: >-
+          Declare this caller an automatic ASAN presubmit gate: a run scoped
+          cheaply enough to start on every pull request. Lets host-asan
+          allocate a sandbox test runner outside of nightly.
+        type: boolean
+        default: false
```
Comment (ScottTodd) [inline .github/workflows/setup_multi_arch.yml:103]: Here and elsewhere, focus on what the option does and leave the commentary for other files.

```suggestion
      asan_presubmit:
        description: "Allows the host-asan build variant CI to run on pull_request events"
        type: boolean
        default: false
```

File: .github/workflows/setup_multi_arch.yml
Comment (ScottTodd) [inline .github/workflows/setup_multi_arch.yml:1]: nit: the PR description is difficult to follow, please rephrase.

Specifically:

> Allow host-asan tests to run on pull_request triggers when the build is from an external repo that has narrowed the build graph via build_stages. This ensures only repos that have explicitly optimized for presubmit (e.g., rocm-systems with compiler-runtime + runtime-tests) receive sandbox runners on every PR.
> 
> Satisfies issue #7202 requirement: "Host-ASan tests receive an eligible sandbox runner on the PR trigger."
> 
> ISSUE ID: #7202

"Receive sandbox runners on every PR" is not a first order goal, it's a prerequisite for these workflows to function. The goal is to run host-asan CI on `pull_request` events and that is what this does. Sandbox runners aren't directly related to the code changes here.

Please also use the supplied pull request template (https://github.com/ROCm/TheRock/blob/main/.github/pull_request_template.md) to make review easier.

File: .github/workflows/setup_multi_arch.yml [hunk 4]
```diff
@@ -227,6 +234,7 @@ jobs:
           BUILD_PYTORCH: ${{ inputs.build_pytorch }}
           BUILD_JAX: ${{ inputs.build_jax }}
           BUILD_NATIVE_LINUX: ${{ inputs.build_native_linux }}
+          ASAN_PRESUBMIT: ${{ inputs.asan_presubmit }}
```
Comment (ScottTodd) [inline .github/workflows/setup_multi_arch.yml:237]: This may be okay for now, but I've given the feedback on multiple PRs at this point that we need to generalize options like this. We have multiple build variants:
* ci/release (default)
* asan sanitizer
* tsan sanitizer
* debug symbols
* code coverage
* canary flags

We shouldn't need to add individual options for each variant, there should be a general way to instantiate a CI pipeline and choose what triggers it should be enabled/disabled for.

File: build_tools/github_actions/configure_multi_arch_ci.py (hunk 2 above)
Comment (dileepr1) [reply build_tools/github_actions/configure_multi_arch_ci.py]: Applied, thanks. I kept external_repo and asan_presubmit on the pull_request row.

File: .github/workflows/setup_multi_arch.yml
Comment (dileepr1) [reply .github/workflows/setup_multi_arch.yml:1]: Rewritten as directed.

File: .github/workflows/setup_multi_arch.yml (hunk 4 above)
Comment (dileepr1) [reply .github/workflows/setup_multi_arch.yml:237]: Thanks, will keep it in consideration going forward.

Comment (ScottTodd) [reply .github/workflows/setup_multi_arch.yml:237]: @geomin12 @jayhawk-commits how do you both feel here? I keep going back and forth on "good enough for now" or "blocking" for this.

As soon as one option like this gets added others may follow, and each repository using it will need to adapt either way. It feels like we're slowly adding complexity over time to CI setup scripts.

Possible approaches:
1. workflow inputs like:
    ```yml
      # rocm-libraries
      with:
        build_variant: asan
        pull_request_mode: automatic
    
      # TheRock
      with:
        build_variant: asan
        pull_request_mode: opt-in
        pull_request_opt_in_label: "ci:asan"
    ```
1. config yaml files for https://github.com/ROCm/TheRock/blob/main/build_tools/github_actions/configure_multi_arch_ci.py like:
    ```yml
      schema_version: 1
    
      variants:
        release:
          events:
            pull_request:
              mode: automatic
              opt_out_label: "ci:skip-release"
            push: {}
            workflow_dispatch: {}
    
        asan:
          events:
            pull_request:
              mode: opt-in
              opt_in_label: "ci:asan"
            schedule: {}
            workflow_dispatch: {}
    
        coverage:
          events:
            workflow_dispatch: {}
    ```
    used like this from the workflows:
    ```yml
      with:
        build_variant: asan
        ci_policy_path: .github/therock-ci.yml
    ```

**Review by ScottTodd — COMMENTED** [review]
(marking as 'reviewed' pending latest comments, I want others to weigh in before I take another look)

File: .github/workflows/setup_multi_arch.yml (hunk 4 above)
Comment (geomin12) [reply .github/workflows/setup_multi_arch.yml:237]: we should block. @raramakr @dileepr1 i understand this is the easiest way to land host-asan for external repos (and fastest as ASAN is important), but Scott is right! we should make this feature reusable, expandable and understandable to other scenarios too. 

these random inputs become tribal knowledge, and it's hard to track as we add more and more "one-off" needed features

for example: i'm doing a ton of re-work to cleanup these "one-off" fixes that have accumulated in https://github.com/ROCm/TheRock/pull/8322, as all of us are losing track of which architecture runs where (is it submodule bump? postsubmit? presubmit?)

we're working on maturing our labelling system so engineers can specify everything they need: https://github.com/ROCm/TheRock/issues/7375, maybe we can start from there? where we automatically add ci:host-asan labels to rocm-libraries / rocm-systems PRs as a quick and clean solution?

and as a permanent solution, we can expand `amdgpu_family_matrix.py` (or somewhere) to make it very clear that: "on presubmit, we run host-asan" or brainstorm to make it clear for everyone (and when we add postsubmit, later features, it'll be easy to add then rather than more inputs)

**Review by geomin12 — COMMENTED** [review]
comment regarding this feature. let's re-work this so it's expandable for later use-cases (using asan for postsubmit should **not** require another input)

as mentioned, all these little fast easy inputs makes it hard to track. will we remember this input 6 months later? (because for me, i definitely will not)

**Comment by dileepr1** [comment]
> comment regarding this feature. let's re-work this so it's expandable for later use-cases (using asan for postsubmit should **not** require another input)
> 
> as mentioned, all these little fast easy inputs makes it hard to track. will we remember this input 6 months later? (because for me, i definitely will not)

@geomin12 @ScottTodd based on the above comments, I've got an prototype in https://github.com/ROCm/TheRock/pull/8455 - could you check if that would be good?
